src/V3Randomize.cpp

```diff
@@ -0,0 +1,280 @@
+                 memberp = VN_CAST(memberp->nextp(), MemberDType)) {
+                auto* randp = newRandStmtsp(fl, stmtsp ? varrefp->cloneTree(false) : varrefp,
+                                            offset, memberp);
+                if (stmtsp)
+                    stmtsp->addNext(randp);
+                else
+                    stmtsp = randp;
+            }
+            return stmtsp;
+        } else {
+            AstNodeMath* valp;
+            if (auto* enumDtp = VN_CAST(memberp ? memberp->subDTypep()->subDTypep()
+                                                : varrefp->dtypep()->subDTypep(),
+                                        EnumDType)) {
+                size_t itemCount = 0;  // Number of enum items
+                for (auto* itemp = enumDtp->itemsp(); itemp;
+                     itemp = VN_CAST(itemp->nextp(), EnumItem))
+                    itemCount++;
+                AstVarRef* tabRefp = new AstVarRef(fl, enumValueTabp(enumDtp), VAccess::READ);
+                tabRefp->classOrPackagep(v3Global.rootp()->dollarUnitPkgAddp());
+                auto* randp = new AstRand(fl, nullptr, false);
+                auto* moddivp = new AstModDiv(fl, randp, new AstConst(fl, itemCount));
+                randp->dtypep(varrefp->findBasicDType(AstBasicDTypeKwd::UINT32));
+                moddivp->dtypep(enumDtp);
+                valp = new AstArraySel(fl, tabRefp, moddivp);
+            } else {
+                valp = new AstRand(fl, nullptr, false);
+                valp->dtypep(memberp ? memberp->dtypep() : varrefp->varp()->dtypep());
+            }
+            return new AstAssign(fl,
+                                 new AstSel(fl, varrefp, offset + (memberp ? memberp->lsb() : 0),
+                                            memberp ? memberp->width() : varrefp->width()),
+                                 valp);
+        }
+    }
+
+    // VISITORS
+    virtual void visit(AstClass* nodep) override {
+        iterateChildren(nodep);
+        if (!nodep->user1()) return;  // Doesn't need randomize, or already processed
+        UINFO(9, "Define randomize() for " << nodep << endl);
+        auto* funcp = V3Randomize::newRandomizeFunc(nodep);
+        auto* fvarp = VN_CAST(funcp->fvarp(), Var);
+        funcp->addStmtsp(new AstAssign(
+            nodep->fileline(), new AstVarRef(nodep->fileline(), fvarp, VAccess::WRITE),
+            new AstConst(nodep->fileline(), AstConst::WidthedValue(), 32, 1)));
+        for (auto* classp = nodep; classp;
+             classp = classp->extendsp() ? classp->extendsp()->classp() : nullptr) {
+            for (auto* memberp = classp->stmtsp(); memberp; memberp = memberp->nextp()) {
+                auto* memberVarp = VN_CAST(memberp, Var);
+                if (!memberVarp || !memberVarp->isRand()) continue;
+                if (VN_IS(memberp->dtypep()->skipRefp(), BasicDType)
+                    || VN_IS(memberp->dtypep()->skipRefp(), StructDType)) {
```

Comment:
skipref can be slow so call once.  More to cleanup below this

```suggestion
                auto* memberDtp = memberp->dtypep()->skipRefp();
                if (VN_IS(memberDtp, BasicDType)
                    || VN_IS(memberDtp, StructDType)) {
```